lang/cs/Org.Apache.REEF.Wake/Time/Time.cs

```diff
@@ -50,33 +55,21 @@ public override bool Equals(object obj)
             {
                 return true;
             }
-            Time other = obj as Time;
+            
+            // Note: This allows for the very strange semantic where
+            // two different subclasses might return true for Equals.
+            var other = obj as Time;
```

Comment:
Also, `Time` really should be an interface. That way, we don't ever run into this issue. Not all implementations of `Time` will be using `TimeStamp` to store the time.